### DISCUSSION TITLE
Getting un expected error message saying msg = "Max cycle reached"

### DISCUSSION THREAD

**Discussion by nagarajtrivedi** [hyperjumptech/grule-rule-engine#201] [General]
Hi all, while executing the rule I am getting an error message as "Max cycle reached"

I will give a brief information on this.

- II have downloaded the the rule engine from github.com/hyperjumtech/grule-rule-engine/
- Built and executed it and it was working properly
- Then I have added another file Hello_test.go under the examples directory
- Actually this file is copied from the file TutorialExample_test.go to the destination file Hello_test.go
- Both of them are existing
- I have changed the fact names, rule name  and done few other modifications in the newly added file Hello_test.go
- When I ran the command make test-short it executed all the rules including the rule from new file added, Hello_test.go
- But I got one error message mentioned below saying "Max cycle reached". Please go through below error
- It also suggested that I need to adjust the maximum cycle using GruleEngine.MaxCycle variable.
- I have done this changes to set MaxCycle value to 10000. But even after that I am getting below error. Let me know what could be the problem

=== RUN   TestHello
time="2021-06-14T16:13:57+05:30" level=error msg="Max cycle reached" lib=grule-rule-engine package=engine
Let say "Hello Nagaraj"
--- FAIL: TestHello (0.03s)
    Hello_test.go:75:
                Error Trace:    Hello_test.go:75
                Error:          Received unexpected error:
                                the GruleEngine successfully selected rule candidate for execution after 5000 cycles, this could possibly caused by rule entry(s) that keep added into execution pool but when executed it does not change any data in context. Please evaluate your rule entries "When" and "Then" scope. You can adjust the maximum cycle using GruleEngine.MaxCycle variable
                Test:           TestHello

**Comment by jinagamvasubabu** [comment]
can you paste the content of the new test file, so we can take a look at it

**Reply by nagarajtrivedi** [reply]
Sure. Please find it below
package examples

import (
        "fmt"
        "github.com/hyperjumptech/grule-rule-engine/ast"
        "github.com/hyperjumptech/grule-rule-engine/builder"
        "github.com/hyperjumptech/grule-rule-engine/engine"
        "github.com/hyperjumptech/grule-rule-engine/pkg"
        "github.com/stretchr/testify/assert"
        "testing"
        "time"
)

type YourFact struct {
        StringAttribute  string
        BooleanAttribute bool
        FloatAttribute   float64
        TimeAttribute    time.Time
        IntAttribute     int64
        WhatToSay        string
}

func (mf *YourFact) ReadWhatToSay(sentence string) string {
        return fmt.Sprintf("Let say \"%s\"", sentence)
}

func TestHello(t *testing.T) {
        yourFact := &YourFact{
                StringAttribute:  "Nagaraj",
                BooleanAttribute: false,
                FloatAttribute:   1.234,
                TimeAttribute:    time.Now(),
                IntAttribute:     254,
        }
 dataCtx := ast.NewDataContext()
        err := dataCtx.Add("YF", yourFact)
        assert.NoError(t, err)

        // Prepare knowledgebase library and load it with our rule.
        knowledgeLibrary := ast.NewKnowledgeLibrary()
        ruleBuilder := builder.NewRuleBuilder(knowledgeLibrary)

        drls := `
rule ChkValues "Check Nagaraj" salience 99 {
    when
        YF.IntAttribute == 254 && YF.BooleanAttribute == false && YF.StringAttribute == "Nagaraj" && YF.FloatAttribute == 1.234
    then
        YF.WhatToSay = YF.ReadWhatToSay("Hello Nagaraj");
                Retract("CheckValues");
}
`
        byteArr := pkg.NewBytesResource([]byte(drls))
        err = ruleBuilder.BuildRuleFromResource("HelloNagaraj", "0.0.1", byteArr)
        assert.NoError(t, err)

        knowledgeBase := knowledgeLibrary.NewKnowledgeBaseInstance("HelloNagaraj", "0.0.1")

        engine := engine.NewGruleEngine()
        /*engine.MaxCycle = 10000 */
        err = engine.Execute(dataCtx, knowledgeBase)
        assert.NoError(t, err)
        assert.Equal(t, "Let say \"Hello Nagaraj\"", yourFact.WhatToSay)
        println(yourFact.WhatToSay)
}

**Reply by newm4n** [reply]
The code snippet you provided is hard to read.
Could you paste it at https://pastebin.com/ ?

**Reply by nagarajtrivedi** [reply]
Hi, I have pasted the code in the link https://pastebin.com/?
Please have a look at it.
It is exactly similar to the file TutorialExample_test.go from the the directory examaple or examples in the rule engine you have created in the github.
I have only changed names of the data structures, few other statements and function names
MyFact is changed to YourFact
myFact is changed to yourFact
MF is changed to YF
The rule name CheckValues is changed to ChkValues
Check Default is changed to Check Nagaraj
The message Hello Grule is chaged to Hello Nagaraj

Let me know at the earliest what could be the problem.

Thanks in advance

Nagaraj Trivedi

**Comment by jinagamvasubabu** [comment]
@nagarajtrivedi, couple of issues identified in your test file
1. `return fmt.Sprintf("Let say "%s"", sentence)` - having syntax issue, you should escape quotes
2. `Retract("CheckValues");` - you are retracting wrong rule
3. `assert.Equal(t, "Let say \"Hello Nagaraj\"", yourFact.WhatToSay)` - assert on wrong value

Coming to maxCycle reached error, its happening because it will try to find the right rule and grule engine not able to find even after 5000 cycles. 

I have updated your test file and please find the correct one below:
```go
package examples

import (
	"fmt"
	"github.com/hyperjumptech/grule-rule-engine/ast"
	"github.com/hyperjumptech/grule-rule-engine/builder"
	"github.com/hyperjumptech/grule-rule-engine/engine"
	"github.com/hyperjumptech/grule-rule-engine/pkg"
	"github.com/stretchr/testify/assert"
	"testing"
	"time"
)

type YourFact struct {
	StringAttribute  string
	BooleanAttribute bool
	FloatAttribute   float64
	TimeAttribute    time.Time
	IntAttribute     int64
	WhatToSay        string
}

func (mf *YourFact) ReadWhatToSay(sentence string) string {
	return fmt.Sprintf("Let say \"%s\"", sentence)
}

func TestHello(t *testing.T) {
	yourFact := &YourFact{
		StringAttribute:  "Nagaraj",
		BooleanAttribute: false,
		FloatAttribute:   1.234,
		TimeAttribute:    time.Now(),
		IntAttribute:     254,
	}
	dataCtx := ast.NewDataContext()
	err := dataCtx.Add("YF", yourFact)
	assert.NoError(t, err)

	// Prepare knowledgebase library and load it with our rule.
	knowledgeLibrary := ast.NewKnowledgeLibrary()
	ruleBuilder := builder.NewRuleBuilder(knowledgeLibrary)

	drls := `
		rule ChkValues "Check Nagaraj" {
			when
			YF.IntAttribute == 254 && YF.BooleanAttribute == false && YF.StringAttribute == "Nagaraj" && YF.FloatAttribute == 1.234
			then
			YF.WhatToSay = YF.ReadWhatToSay("Hello Nagaraj");
			Retract("ChkValues");
      }
`
	byteArr := pkg.NewBytesResource([]byte(drls))
	err = ruleBuilder.BuildRuleFromResource("HelloNagaraj", "0.0.1", byteArr)
	assert.NoError(t, err)

	knowledgeBase := knowledgeLibrary.NewKnowledgeBaseInstance("HelloNagaraj", "0.0.1")

	engine := engine.NewGruleEngine()
	/*engine.MaxCycle = 10000 */
	err = engine.Execute(dataCtx, knowledgeBase)
	assert.NoError(t, err)
	assert.Equal(t, "Let say \"Hello Nagaraj\"", yourFact.WhatToSay)
	println(yourFact.WhatToSay)
}

```

**Reply by nagarajtrivedi** [reply]
It worked. Thanks for your timely help Vasu.

Thanks and Regards

Nagaraj Trivedi
